npm: add module

mirkolenz opened the pull request:
### Description

Add module for configuring `.npmrc` via home manager. Inspired by the [NixOS module](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/programs/npm.nix).

I'm not sure if the tests are set up correctly. Since the package is a nested attrset, I didn't use a stub, but set the package to null instead. Please let me know if there's a better way to handle this.

<!--

Please provide a brief description of your change.

-->

### Checklist

<!--

Please go through the following checklist before opening a non-WIP
pull-request.

Also make sure to read the guidelines found at

  https://nix-community.github.io/home-manager/#sec-guidelines

-->

- [x] Change is backwards compatible.

- [x] Code formatted with `nix fmt` or
      `nix-shell -p treefmt nixfmt deadnix keep-sorted --run treefmt`.

- [x] Code tested through `nix run .#tests -- test-all` or
      `nix-shell --pure tests -A run.all`.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

  ```
  {component}: {description}

  {long description}
  ```

  See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/a51598236f23c89e59ee77eb8e0614358b0e896c/modules/programs/lesspipe.nix#L11).
  - [x] Generate a news entry. See [News](https://nix-community.github.io/home-manager/index.xhtml#sec-news)
  - [x] Basic tests added. See [Tests](https://nix-community.github.io/home-manager/index.xhtml#sec-tests)

- If this PR adds an exciting new feature or contains a breaking change.
  - [ ] Generate a news entry. See [News](https://nix-community.github.io/home-manager/index.xhtml#sec-news)

0xda157 commented on modules/programs/npm.nix:
```diff
@@ -0,0 +1,53 @@
+      };
+      sessionVariables = lib.mkIf (cfg.npmrc != "") {
+        NPM_CONFIG_USERCONFIG = "${config.home.homeDirectory}/.npmrc";
+      };
```
this should use `~/.config/npm/npmrc` when `home.preferXdgDirectories` is enabled

mirkolenz commented:
Ok I think I found an elegant solution: I created a custom ini generation function which automatically appends square brackets to list values. I renamed the option from `npmrc` to `settings` with a proper type. I updated the tests as well by adding an array-based entry.

One thing I changed as well compared to the last iteration: The package option now installs the `nodejs` package because at least on my system, using `pkgs.nodePackages.npm` or `pkgs.nodejs.pkgs.npm` resulted in a full rebuild of the nodejs package (which is expensive). It also seems to be the case that `pkgs.nodePackages_xx` is no longer available and just not updated in the upstream NixOS module.

Let me know what you think 😄
